src/zarr/v3/codecs/batched_pipeline.py

```diff
@@ -0,0 +1,283 @@
+    ) -> None:
+        assert self.supports_partial_encode
+        assert isinstance(self.array_bytes_codec, ArrayBytesCodecPartialEncodeMixin)
+        await self.array_bytes_codec.encode_partial_batched(batch_info, runtime_configuration)
+
+    def compute_encoded_size(self, byte_length: int, array_spec: ArraySpec) -> int:
+        for codec in self:
+            byte_length = codec.compute_encoded_size(byte_length, array_spec)
+            array_spec = codec.resolve_metadata(array_spec)
+        return byte_length
+
+    async def write_batched(
+        self,
+        batch_info: Iterable[Tuple[StorePath, ArraySpec, SliceSelection, SliceSelection]],
+        value: np.ndarray,
+        runtime_configuration: RuntimeConfiguration,
+    ) -> None:
+        if self.supports_partial_encode:
+            await self.encode_partial_batched(
+                [
+                    (store_path, value[out_selection], chunk_selection, chunk_spec)
+                    for store_path, chunk_spec, chunk_selection, out_selection in batch_info
+                ],
+                runtime_configuration,
+            )
+
+        else:
+            # Read existing bytes if not total slice
+            async def _read_key(store_path: Optional[StorePath]) -> Optional[BytesLike]:
+                if store_path is None:
+                    return None
+                return await store_path.get()
+
+            chunk_bytes_batch = await concurrent_map(
+                [
+                    (None if is_total_slice(chunk_selection, chunk_spec.shape) else store_path,)
+                    for store_path, chunk_spec, chunk_selection, _ in batch_info
+                ],
+                _read_key,
+                runtime_configuration.concurrency,
+            )
+            chunk_array_batch = await self.decode_batched(
+                [
+                    (chunk_bytes, chunk_spec)
+                    for chunk_bytes, (_, chunk_spec, _, _) in zip(chunk_bytes_batch, batch_info)
+                ],
+                runtime_configuration,
+            )
+
+            def _merge_chunk_array(
+                existing_chunk_array: Optional[np.ndarray],
+                new_chunk_array_slice: np.ndarray,
+                chunk_spec: ArraySpec,
+                chunk_selection: SliceSelection,
+            ) -> np.ndarray:
+                if is_total_slice(chunk_selection, chunk_spec.shape):
+                    return new_chunk_array_slice
+                if existing_chunk_array is None:
+                    chunk_array = np.empty(
+                        chunk_spec.shape,
+                        dtype=chunk_spec.dtype,
+                    )
+                    chunk_array.fill(chunk_spec.fill_value)
+                else:
+                    chunk_array = existing_chunk_array.copy()  # make a writable copy
+                chunk_array[chunk_selection] = new_chunk_array_slice
+                return chunk_array
+
+            chunk_array_batch = [
```

Comment:
Any way to parallelize these portions between reading and decoding and subsequent encoding?

Reply:
We could certainly parallelize that with `concurrent_map`. Not sure how that would be parallelized using the GPU.